SplitFiles/customer.c

```diff
@@ -0,0 +1,168 @@
+//////////////////////////////////////////////////////////
+//
+// Customers
+//
+//////////////////////////////////////////////////////////
+
+// Can be compiled with:
+// zcc +zx -vn -startup=1 -clib=sdcc_iy -D_TEST_CUSTOMER bread.c util.c customer.c -o customer -create-app
+
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "customer.h"
+#include "bread.h"
+#include "util.h"
+
+void redraw_customers(CustomerBase *base) {
+    unsigned char ypos = 4;
+    Customer* lastCustomer = base->rootCustomer;
+    while (lastCustomer->nextCustomer != NULL) {
+        lastCustomer = lastCustomer->nextCustomer;
+        printf(PRINTAT"\x12%c""%-12s", ypos, lastCustomer->breadOrder->desc);
+        ++ypos;
+    }
+    for (; ypos < 4 + MAX_CUSTOMER_COUNT; ++ypos) {
+        printf(PRINTAT"\x12%c""%-12s", ypos, "");
+    }
+}
+
+void customer_func(GameComponent* customers, GameParameters* params) {
+    CustomerBase* base = (CustomerBase*)(customers->ptr);
+    Customer* newCustomer = (Customer*)NULL;
+    int redraw = 0;
+
+    Customer* thisCustomer;
+    Customer* lastCustomer;
+
+    if (params->messageAddress == 20) {
+        //Someone sent toast!
+        BreadState* bread = (BreadState*)params->message;
+        thisCustomer = base->rootCustomer->nextCustomer;
+        lastCustomer = base->rootCustomer;
+        while (thisCustomer != NULL) {
+            if (bread->type->letter == thisCustomer->breadOrder->letter) {
+                //We have a match
+                params->score += bread->type->cost * 2 //SHOCKING markup
+                         + 100 - abs(bread->toastedness - 100) //100 toastedness gets 100 bonus points
+                         + thisCustomer->ticksLeft;  //Speedy toast = points!
+                (params->slices) += 1;
+                params->messageAddress = 0;
+                params->message = NULL;
+                params->messageSourceAddress = NULL;
+                //Give back memory
+                free(bread);
+                //Remove customer
+                base->customerCount -= 1;
+                lastCustomer->nextCustomer = thisCustomer->nextCustomer;
+                free(thisCustomer);
```

Comment:
Need an end-of-game cleanup function?  Eg to recursively free any allocated customers?

Reply:
Right now the machine basically crashes at end of game so that's not an issue!!

At some point we'll need to do exactly that though.  All the mallocs will need a matched off free somewhere (unless we have something like the BreadBin set up once on load and just reuse it through multiple playings)